misc/dbt-materialize/dbt/include/materialize/macros/is_cluster_ready.sql

```diff
@@ -0,0 +1,140 @@
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+{% macro is_cluster_ready(cluster=target.cluster|default(none)) %}
+
+-- Check if the cluster argument was provided or exists in the target profile
+{% if cluster is none %}
+    {{ exceptions.CompilationError("No cluster specified and no default cluster found in target profile. " ~ current_target_name) }}
+{% endif %}
+
+{{ log("Checking pending objects for cluster " ~ cluster, info=True) }}
+
+{% set cluster_exists %}
+WITH total_replicas AS (
+    SELECT cluster_id, count(*) AS replicas
+    FROM mz_cluster_replicas
+    GROUP BY cluster_id
+)
+
+SELECT COALESCE(replicas, 0) AS replicas
+FROM mz_clusters
+LEFT JOIN total_replicas ON mz_clusters.id = cluster_id
+WHERE mz_clusters.name = lower('{{ cluster }}')
+{%- endset -%}
+
+{%- set result = run_query(cluster_exists) %}
+{%- if execute -%}
+    {%- if result and result.rows|length > 0 -%}
+        {%- set replicas_value = result.rows[0][0] %}
+        {%- if replicas_value is none or replicas_value == 0 -%}
+            {{ log("Cluster " ~ cluster ~ " has no running replicas", info=true) }}
+            {{ return(false) }}
+        {%- endif -%}
+    {%- else -%}
+        {{ log("Cluster " ~ cluster ~ " does not exist", info=true) }}
+        {{ return(false) }}
+    {%- endif -%}
+{%- endif -%}
+
+{%- set check_pending_objects_sql %}
+WITH dataflows AS (
+    SELECT mz_indexes.id, mz_indexes.name, 'index' AS type
+    FROM mz_indexes
+    JOIN mz_clusters ON mz_indexes.cluster_id = mz_clusters.id
+    WHERE mz_clusters.name = lower('{{ cluster }}')
+
+    UNION ALL
+
+    SELECT mz_materialized_views.id, mz_materialized_views.name, 'materialized-view' AS type
+    FROM mz_materialized_views
+    JOIN mz_clusters ON mz_materialized_views.cluster_id = mz_clusters.id
+    WHERE mz_clusters.name = lower('{{ cluster }}')
+),
+
+ready_dataflows AS (
+    SELECT id, name, type
+    FROM dataflows AS d
+    JOIN mz_internal.mz_compute_hydration_statuses AS h ON (h.object_id = d.id)
+    LEFT JOIN mz_internal.mz_materialization_lag AS l ON (l.object_id = d.id)
+    WHERE h.hydrated AND (l.local_lag <= '1s' OR l.local_lag IS NULL)
+),
+
+pending_dataflows AS (
+    SELECT id, name, type
+    FROM dataflows d
+
+    EXCEPT
+
+    SELECT id, name, type
+    FROM ready_dataflows r
+
+),
+
+pending_sources AS (
+    SELECT mz_sources.id, mz_sources.name, 'source' AS type
+    FROM mz_sources
+    JOIN mz_clusters ON mz_clusters.id = mz_sources.cluster_id
+    JOIN mz_internal.mz_source_statistics s ON mz_sources.id = s.id
+    WHERE mz_clusters.name = lower('{{ cluster }}')
+        AND snapshot_committed IS FALSE
```

Comment:
Note that `snapshot_committed` will be `true` after the first time the source has hydrated, and remains `true` across replica restarts. If you want to check whether a source has hydrated on the current replica incarnation, you can use `rehydration_latency` instead.